docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,372 @@
+
+## Rule priorities
+
+When several rules are defined for one row, the most granular is chosen for this row. More specifically, the rule priority is: Index > Partition > Table > Database > Default.
+
+For example:
+
+1. At the beginning, all rows are placed based on the default placement rules.
+2. When a placement rule is added on table `t`, all data on `t` is placed based on the rule of `t`.
+3. When a placement rule is added on partition `p0` of `t`, all data on `p0` is placed based on the rule of `p0`, but other partitions stay still.
+4. When the placement rule on `p0` is removed, data on `p0` is placed based on the rule of `t`, just like other partitions.
+
+Rules priorities should be checked once a placement is added, altered, or dropped.
+
+Rule priorities can be implemented by properties `index` and `override`. `override` is alway enabled, and `index` stands for the priority. Specifically, `index` values can be like this:
+
+* `index` of default placement rules is 0
+* `index` of database placement rules is 1
+* `index` of table placement rules is 2
+* `index` of partition placement rules is 3
+* `index` of index placement rules is 4
+
+In such a way, the most granular rule always works.
+
+When multiple placement rules are defined on the same granular object, the new one always overwrites the old one. To achieve this, property `group_id` in placement rules can be used to identify the database object.
+
+For example, `group_id` can be in such a format:
+
+* `group_id` of a database is the database id
+* `group_id` of a table is the table id
+* `group_id` of a partition is the partition id
+* `group_id` of an unpartitioned index is the concatenation of table id and index id, e.g. `100-1`
+* `group_id` of a partitioned index is the concatenation of partition id and index id
+
+## DDL management
+
+DDL on database objects can also affect placement rules.
+
+### DDL on tables
+
+Once a table is created, it follows the default placement rule.
+
+Once a table is dropped, the placement rules on it cannot be dropped immediately, because the table can be recovered by `FLASHBACK` or `RECOVER` statement before GC collects the data. Related placement rules should be kept temporarily and will be removed after GC lifetime.
+
+When it’s time to remove all related placement rules, not only those rules defined on the table should be removed, but also rules defined on its partitions and indexes.
+
+Once the table is truncated, the placement rules on the table should be updated. Since table id is updated after truncating, key range is also changed.
+
+Since the table can be recovered later by `FLASHBACK` statement, a snapshot of the original placement rules must be saved temporarily. After recovering, the table name is changed, but the table id is the original one, so the snapshot of the original placement rules can be recovered directly.
```

Comment:
How will this be done? It seems complicated IMHO.

Reply:
Addressed. It's checked in GC worker.